circuit-overlay/src/commonMain/kotlin/com/slack/circuit/overlay/ContentWithOverlays.kt

```diff
@@ -18,17 +24,48 @@ import androidx.compose.ui.Modifier
  * @param overlayHost the [OverlayHost] to use for managing overlays.
  * @param content The regular content to render. Any overlays will be rendered over them.
  */
+@OptIn(ExperimentalAnimationApi::class)
 @Composable
 public fun ContentWithOverlays(
   modifier: Modifier = Modifier,
   overlayHost: OverlayHost = rememberOverlayHost(),
   content: @Composable () -> Unit
 ) {
   CompositionLocalProvider(LocalOverlayHost provides overlayHost) {
-    val overlayHostData by rememberUpdatedState(overlayHost.currentOverlayData)
     Box(modifier) {
       content()
-      key(overlayHostData) { overlayHostData?.let { data -> data.overlay.Content(data::finish) } }
+      val transition = updateTransition(overlayHost.currentOverlayData, label = null)
+      val enterTransition =
+        remember(transition.targetState) {
+          (transition.targetState?.overlay as? AnimatedOverlay)?.enterTransition
+            ?: EnterTransition.None
+        }
+      val exitTransition =
+        remember(transition.currentState) {
+          (transition.currentState?.overlay as? AnimatedOverlay)?.exitTransition
+            ?: ExitTransition.None
+        }
+      transition.AnimatedVisibility(
+        visible = { it != null },
+        modifier = Modifier.fillMaxSize(),
+        enter = enterTransition,
+        exit = exitTransition,
+      ) {
+        // Compute the current overlay data from the transition.
+        // If transitioning from null to not null, we want to use the targetState. And vice-versa,
+        // if transitioning from not null to null, we want to use the currentState.
+        val overlayData by remember {
+          derivedStateOf { transition.currentState ?: transition.targetState }
```

Comment:
The `derivedStateOf` isn't doing much here. You can remove it and just inline the null coalesce. Also, swapping the content is making me think that an `AnimatedContent` is a better fit here.

Reply:
Let me take another shot using `AnimatedContent`. I ran into some issues with it in my first draft, but it definitely makes the transition spec calculation nicer...

On the other hand, maybe `AnimatedContent` should just observe the `Transition`'s state and animate itself entirely...

```
@Composable
public abstract fun Transition< EnterExitState>(navigator: OverlayNavigator<Result>)
```
I don't love how we have to change the enter/exit inputs with each state change in the current design.